zerver/tests/test_event_queue.py

```diff
@@ -86,3 +92,65 @@ def test_enqueue_notifications(self):
             always_push_notify=True, idle=False)
         self.assertTrue(email_notice is None)
         self.assertTrue(mobile_notice is not None)
+
+    def tornado_call(self, view_func, user_profile, post_data):
+        # type: (Callable[[HttpRequest, UserProfile], HttpResponse], UserProfile, Dict[str, Any]) -> HttpResponse
+        request = POSTRequestMock(post_data, user_profile)
+        return view_func(request, user_profile)
+
+    def test_end_to_end_missedmessage_hook(self):
+        # type: () -> None
+        """Tests what arguments missedmessage_hook passes into maybe_enqueue_notifications.
+        Combined with the previous test, this ensures that the missedmessage_hook is correct"""
+        user_profile = self.example_user('hamlet')
+        email = user_profile.email
+        self.login(email)
+
+        result = self.tornado_call(get_events_backend, user_profile,
+                                   {"apply_markdown": ujson.dumps(True),
+                                    "event_types": ujson.dumps(["message"]),
+                                    "user_client": "website",
+                                    "dont_block": ujson.dumps(True),
+                                    })
+        self.assert_json_success(result)
+        queue_id = ujson.loads(result.content)["queue_id"]
+        client_descriptor = get_client_descriptor(queue_id)
+
+        with mock.patch("zerver.tornado.event_queue.maybe_enqueue_notifications") as mock_enqueue:
+            # To test the missed_message hook, we first need to send a message
+            msg_id = self.send_message(self.example_email("iago"), "Denmark", Recipient.STREAM)
+
+            # Verify that nothing happens if you call it as not the last client
```

Comment:
Can you explain not-the-last-client?

Reply:
This is the top of the function:

```
def missedmessage_hook(user_profile_id, client, last_for_client):                                                   
    # type: (int, ClientDescriptor, bool) -> None                                                                   
    # Only process missedmessage hook when the last queue for a                                                     
    # client has been garbage collected                                                                             
    if not last_for_client:                                                                                         
        return                                                                                                      
```

Do you think we need to also explain that in the test?

Reply:
definitely